File: engine/src/main/java/de/hanno/hpengine/engine/graphics/query/GLTimerQuery.java
package de.hanno.hpengine.engine.graphics.query;

import de.hanno.hpengine.engine.backend.OpenGl;
import de.hanno.hpengine.engine.graphics.GpuContext;
import kotlin.Unit;
import org.lwjgl.opengl.GL15;

import static org.lwjgl.opengl.GL15.GL_QUERY_RESULT;
import static org.lwjgl.opengl.GL15.glGenQueries;
import static org.lwjgl.opengl.GL33.*;

public class GLTimerQuery implements GLQuery<Float> {

    private GpuContext<OpenGl> gpuContext;

    private final int start;
    private final int end;
    private volatile boolean finished = false;
    private boolean started;

    public GLTimerQuery(GpuContext gpuContext) {
        this.gpuContext = gpuContext;
        start = glGenQuery();
        end = glGenQuery();
    }

    private int glGenQuery() {
        return gpuContext.invoke(GL15::glGenQueries);
    }

    @Override
    public GLTimerQuery begin() {
        finished = false;
        gpuContext.invoke(() -> {
            glQueryCounter(start, GL_TIMESTAMP);
            return Unit.INSTANCE;
        });
        started = true;
        return this;
    }

    @Override
    public void end() {
        if(!started) {
            throw new IllegalStateException("Don't end a query before it was started!");
        }
        gpuContext.invoke(() -> {
            glQueryCounter(end, GL_TIMESTAMP);
            return Unit.INSTANCE;
        });
        finished = true;
    }


    public long getTimeTaken() {
        return getEndTime() - getStartTime();
    }

    @Override
    public int getQueryToWaitFor() {
        return start;
    }

    public long getStartTime() {
        return gpuContext.invoke(() -> glGetQueryObjectui64(start, GL_QUERY_RESULT));
    }

    public long getEndTime() {
        return gpuContext.invoke(() -> glGetQueryObjectui64(end, GL_QUERY_RESULT));
    }

    @Override
    public Float getResult() {
        if(!finished) { throw new IllegalStateException("Don't query result before query is finished!"); }
        while(!resultsAvailable(gpuContext)) {
        }
        return getTimeTaken() / 1000000f;
    }
}

File: engine/src/main/kotlin/de/hanno/hpengine/engine/graphics/query/GLTimerQuery.kt
package de.hanno.hpengine.engine.graphics.query

import de.hanno.hpengine.engine.graphics.GpuContext
import org.lwjgl.opengl.GL15
import org.lwjgl.opengl.GL33

class GLTimerQuery(gpuContext: GpuContext<*>) : GLQuery<Float?> {
    private val gpuContext: GpuContext<*>
    override val queryToWaitFor: Int
    private val end: Int

    @Volatile
    private var finished = false
    private var started = false
    private fun glGenQuery(): Int {
        return gpuContext.invoke { GL15.glGenQueries() }
    }

    override fun begin(): GLTimerQuery? {
        finished = false
        gpuContext.invoke {
            GL33.glQueryCounter(queryToWaitFor, GL33.GL_TIMESTAMP)
            Unit
        }
        started = true
        return this
    }

    override fun end() {
        check(started) { "Don't end a query before it was started!" }
        gpuContext.invoke {
            GL33.glQueryCounter(end, GL33.GL_TIMESTAMP)
            Unit
        }
        finished = true
    }

    val timeTaken: Long
        get() = endTime - startTime
    val startTime: Long
        get() = gpuContext.invoke { GL33.glGetQueryObjectui64(queryToWaitFor, GL15.GL_QUERY_RESULT) }
    val endTime: Long
        get() = gpuContext.invoke { GL33.glGetQueryObjectui64(end, GL15.GL_QUERY_RESULT) }
    override val result: Float
        get() {
            check(finished) { "Don't query result before query is finished!" }
            while (!resultsAvailable(gpuContext)) {
            }
            return timeTaken / 1000000f
        }

    init {
        this.gpuContext = gpuContext
        queryToWaitFor = glGenQuery()
        end = glGenQuery()
    }
}